services/QuillLMS/client/app/bundles/Evidence/libs/stringFormatting.test.ts

```diff
@@ -1,34 +1,79 @@
-import { highlightSpellingGrammar } from "./stringFormatting"
+import { iteratee } from "lodash";
+
+import { highlightGrammar, highlightSpelling } from "./stringFormatting"
 
 
 describe('stringFormatting', () => {
-  describe('highlightSpellingGrammar', () => {
+  describe('highlightSpelling', () => {
     it('basic', () => {
-      const result = highlightSpellingGrammar('lorem ipsum', 'ipsum')
+      const result = highlightSpelling('lorem ipsum', 'ipsum')
       expect(result).toEqual("lorem <b>ipsum</b>")
     });
 
     it('should accept an array for wordsToFormat arg', () => {
-      const result = highlightSpellingGrammar('lorem ipsum foo', ['ipsum', 'foo'])
+      const result = highlightSpelling('lorem ipsum foo', ['ipsum', 'foo'])
       expect(result).toEqual("lorem <b>ipsum</b> <b>foo</b>")
     });
 
     it('should highlight all instances of a matchword', () => {
-      const result = highlightSpellingGrammar(`lorem ipsum lorem`, 'lorem')
+      const result = highlightSpelling(`lorem ipsum lorem`, 'lorem')
       expect(result).toEqual(`<b>lorem</b> ipsum <b>lorem</b>`)
     });
 
     it('should highlight all of a word when a word substring is matched', () => {
-      const result = highlightSpellingGrammar("lorem shouldn't ipsum", 'should')
+      const result = highlightSpelling("lorem shouldn't ipsum", 'should')
       expect(result).toEqual(`lorem <b>shouldn't</b> ipsum`)
     });
     it('should highlight all of a word when a word substring is matched', () => {
-      const result = highlightSpellingGrammar("lorem shouldn't ipsum", 'should')
+      const result = highlightSpelling("lorem shouldn't ipsum", 'should')
       expect(result).toEqual(`lorem <b>shouldn't</b> ipsum`)
     });
     it('should not highlight nonbreaking space character', () => {
-      const result = highlightSpellingGrammar("&nbsp lorem ipsum shouldn't", "&nbsp")
+      const result = highlightSpelling("&nbsp lorem ipsum shouldn't", "&nbsp")
       expect(result).toEqual(`&nbsp lorem ipsum shouldn't`)
     });
   });
+
+  describe('highlightGrammar', () => {
+    it('return the original string input when highlights are empty', () => {
+      const result = highlightGrammar('lorem ipsum', [])
+      expect(result).toEqual('lorem ipsum')
+    });
+
+    it('should bold, given 1 highlight', () => {
+      const result = highlightGrammar('lorem ipsum', [{character: 6, text: "ipsum"}])
+      expect(result).toEqual('lorem <b>ipsum</b>')
+    });
+
+    it('should not bold a phrase before the given character', () => {
+      const result = highlightGrammar('lorem ipsum ipsum', [{character: 8, text: "ipsum"}])
+      expect(result).toEqual('lorem ipsum <b>ipsum</b>')
+    });
+
+    it('should not bold a phrase after the given character', () => {
+      const result = highlightGrammar('lorem ipsum ipsum', [{character: 0, text: "ipsum"}])
+      expect(result).toEqual('lorem ipsum ipsum')
+    });
```

Comment:
Is this really the condition we want here?  I would have thought that this should highlight the first instance of "ipsum" it finds rather than nothing.  What am I missing?

Reply:
Example grammar engine offset payload: https://quill-org.postman.co/workspace/Quill-Team-Workspace~be8cf879-9fa6-4dc5-8544-630985cdf233/request/14211978-14b72239-9722-4049-9ab9-a774f4696afd

Reply:
I think I'm still confused here.  The `it` text here suggests that if the offset is high enough, nothing should be highlighted, but I don't understand why nothing would be highlighted if the offset were _too low_.  Again, I'm probably missing something here, but can't figure out what.

Reply:
(TR and I resolved this over zoom)